### PR TITLE
DD-19 added captcha

### PR DESCRIPTION
**Pull request by YuraStork**

### CODE REVIEW

File: src/api/user/authorize.ts [hunk 1]
```diff
@@ -2,5 +2,5 @@ import { Instance } from 'api/instance'
 
 import { AuthResponse, User2FALoginResponse, UserLoginFormData } from 'types'
 
-export const authorizeUser = async (data: UserLoginFormData) =>
+export const authorizeUser = async (data: UserLoginFormData & { captcha: string }) =>
```
Comment (Andrew01011997) [inline src/api/user/authorize.ts]: create new interface based on UserLoginFormData

File: src/api/user/registration.ts [hunk 2]
```diff
@@ -2,6 +2,6 @@ import { Instance } from 'api/instance'
 
 import { UserRegistrationData } from 'types'
 
-export const registrationUser = async (data: UserRegistrationData) => {
+export const registrationUser = async (data: UserRegistrationData & { captcha: string }) => {
```
Comment (Andrew01011997) [inline src/api/user/registration.ts]: Same as above

File: src/pages/auth/login/index.tsx [hunk 3]
```diff
@@ -24,23 +28,36 @@ import { LoginFileds, initialValues } from './const'
...
-
-  const setAttemptsLeftCount = (count: number) => dispatch(setAttemptsLeftCountAction(count))
-
   const formik = useFormik({
     initialValues,
     validationSchema: loginValidationSchema,
     validateOnBlur: true,
     validateOnChange: true,
-    onSubmit: (data) => dispatch(loginThunk({ ...data, setAttemptsLeftCount }))
+    onSubmit: noopFunction
```
Comment (Andrew01011997) [inline src/pages/auth/login/index.tsx:40]: Why you use `noop` here?

File: src/pages/auth/login/index.tsx [hunk 4]
```diff
@@ -24,23 +28,36 @@ import { LoginFileds, initialValues } from './const'
...
   })
 
+  const onSuccessCallback = (data: AuthResponse) => dispatch(saveUserDataThunk(data))
+  const onErrorCallback = (err: string) => toast.error(err)
```
Comment (Andrew01011997) [inline src/pages/auth/login/index.tsx:44]: Does it make sense?

File: src/pages/auth/login/index.tsx [hunk 5]
```diff
@@ -24,23 +28,36 @@ import { LoginFileds, initialValues } from './const'
...
+  const handleCloseModal = () => dispatch(cancelUser2faAction())
+
+  const setAttemptsLeftCount = useCallback(
+    (count: number) => dispatch(setAttemptsLeftCountAction(count)),
+    [dispatch]
+  )
```
Comment (Andrew01011997) [inline src/pages/auth/login/index.tsx:50]: dispatch is unnecessary dependency here

File: src/pages/auth/login/index.tsx [hunk 6]
```diff
@@ -24,23 +28,36 @@ import { LoginFileds, initialValues } from './const'
...
+
+  const handleSubmit = useCallback(() => {
+    if (!executeRecaptcha) {
+      toast.error(ErrorMessages.INVALID_RECAPTCHA)
+      return
+    }
+
+    dispatch(loginThunk({ ...formik.values, setAttemptsLeftCount, executeRecaptcha }))
```
Comment (Andrew01011997) [inline src/pages/auth/login/index.tsx]: Why you add cb inside dispatch?

File: src/pages/auth/registration/index.tsx [hunk 7]
```diff
@@ -1,38 +1,45 @@
...
+  const { executeRecaptcha } = useGoogleReCaptcha()
+  const dispatch = useAppDispatch()
 
   const formik = useFormik({
     initialValues,
     validationSchema: registrationValidationSchema,
     validateOnBlur: true,
     validateOnChange: true,
-    onSubmit: handleSubmit
+    onSubmit: noopFunction
```
Comment (Andrew01011997) [inline src/pages/auth/registration/index.tsx:31]: Why noop?

File: src/pages/auth/registration/index.tsx [hunk 8]
```diff
@@ -47,7 +54,9 @@ export const RegistrationComponent = () => {
               value={formik.values[field]}
             />
           ))}
-          <AuthButton disabled={!formik.isValid || isLoading}>Sign up</AuthButton>
+          <AuthButton disabled={!formik.isValid || isLoading} onClick={handleSubmit} type='button'>
+            Sign up
+          </AuthButton>
```
Comment (Andrew01011997) [inline src/pages/auth/registration/index.tsx:64]: Can `type="button"` be inside `AuthButton`?

File: src/router.tsx [hunk 9]
```diff
@@ -63,7 +65,14 @@ const setRoutes = (isAuth: boolean) =>
       ]
     : [
         { path: '/', element: <Navigate to='/authorization' /> },
-        { path: '/authorization', element: <AuthPage /> },
+        {
+          path: '/authorization',
+          element: (
+            <GoogleReCaptchaProvider {...reCaptchaConfig}>
```
Comment (Andrew01011997) [inline src/router.tsx]: You don't use any dynamic data here, could this provider be inside AuthPage? or create wrapper component

File: src/store/thunks/user/authorization.thunk.ts [hunk 10]
```diff
@@ -42,13 +42,27 @@ export const updateAuthStatusThunk = createAsyncThunk(
...
-  async ({ setAttemptsLeftCount, ...data }: LoginThunkParams, { dispatch }) => {
+  async (
+    { email, password, setAttemptsLeftCount, executeRecaptcha }: LoginThunkParams,
+    { dispatch }
+  ) => {
     try {
-      const password = cryptoSha256(data.password)
-      const response = await authorizeUser({ email: data.email, password })
+      const hashPassword = cryptoSha256(password)
+      const captcha = await executeRecaptcha('login')
```
Comment (Andrew01011997) [inline src/store/thunks/user/authorization.thunk.ts]: What is the type of `captcha`? Could you pass it as a parameter to this function?

File: src/store/thunks/user/authorization.thunk.ts [hunk 11]
```diff
@@ -42,13 +42,27 @@ export const updateAuthStatusThunk = createAsyncThunk(
...
+      }
+
+      const response = await authorizeUser({
+        email,
+        password: hashPassword,
+        captcha
+      })
 
       if (Object.hasOwn(response.data, 'isUse2FA')) {
         setAttemptsLeftCount((response.data as User2FALoginResponse).attemptsLeftCount)
```
Comment (Andrew01011997) [inline src/store/thunks/user/authorization.thunk.ts]: could you don't use cb inside thunk?

Comment (Andrew01011997) [reply src/store/thunks/user/authorization.thunk.ts]: Variant of better usage is `dispatch(...).unwrap().then(...)`

File: src/store/thunks/user/authorization.thunk.ts [hunk 12]
```diff
@@ -77,10 +91,19 @@ export const saveUserDataThunk = createAsyncThunk(
 
 export const userRegistrationThunk = createAsyncThunk(
   `${USER_SLICE_NAME}/registration-thunk`,
-  async (data: UserRegistrationData, { rejectWithValue }) => {
+  async ({ password, executeRecaptcha, ...data }: RegistrationThunkParams, { rejectWithValue }) => {
     try {
-      const response = await registrationUser(data)
+      const hash_password = cryptoSha256(password)
+      const captcha = await executeRecaptcha('login')
```
Comment (Andrew01011997) [inline src/store/thunks/user/authorization.thunk.ts]: Same as above

File: src/store/thunks/user/authorization.thunk.ts [hunk 13]
```diff
@@ -77,10 +91,19 @@ export const saveUserDataThunk = createAsyncThunk(
...
+
+      if (!captcha) {
+        throw 'error'
+      }
+
+      const response = await registrationUser({ ...data, password: hash_password, captcha })
```
Comment (Andrew01011997) [inline src/store/thunks/user/authorization.thunk.ts]: Please rename `registrationUser` to `createUser`

**Review by Andrew01011997 — CHANGES_REQUESTED** [review]

File: src/pages/auth/login/index.tsx (hunk 3 above)
Comment (YuraStork) [reply src/pages/auth/login/index.tsx:40]: Because onSubmit will not call, and its like a plug

File: src/pages/auth/login/index.tsx (hunk 5 above)
Comment (YuraStork) [reply src/pages/auth/login/index.tsx:50]: Eslint will be swear

File: src/pages/auth/login/index.tsx (hunk 6 above)
Comment (YuraStork) [reply src/pages/auth/login/index.tsx]: Because not writing logic inside this function

File: src/pages/auth/registration/index.tsx (hunk 7 above)
Comment (YuraStork) [reply src/pages/auth/registration/index.tsx:31]: Because onSubmit will not call, and its like a plug

**Review by Andrew01011997 — APPROVED** [review]

**Pull request merged**